### PR TITLE
fix: ensure ES/OS cluster setup for shared usage

### PR DESCRIPTION
**Pull request by gabor-boros**
## Description

The current configuration had several smaller misconfigurations and inconsistent naming, which made it challenging to configure the cluster properly. Also, there were unnecessary configuration steps that I got rid of.

This pull request refactors the way we set up and configure both Open Search and Elasticsearch for shared usage; also

* fixes naming inconsistencies
* reduces the number of config lines we have to include for an instance
* simplifies and fixes Elasticsearch user creation
* fixes https://github.com/openedx/openedx-k8s-harmony/issues/48
* resolves https://github.com/openedx/openedx-k8s-harmony/issues/23

## Dependencies

- https://github.com/overhangio/tutor-forum/pull/41

## Testing

1. Create a cluster
2. Install the harmony helm-chart with ES enabled
3. Create a new user for the instance with the `create_elasticsearch_user` command
4. Install [this version](https://github.com/overhangio/tutor-forum/pull/41) of the tutor-forum plugin
5. Deploy the instance to the cluster
6. Create a new course
7. Go and start some discussions

## Screenshots

Screenshot 2024-07-30 at 19 16 41

## Notes

I have tested that the Elasticsearch cluster is spinning up correctly, the user is created, the instance is deployed with tutor-forum installed. Also, the instance seems to work fine. However, I did not test the Open Search cluster -- though it should work.

### CODE REVIEW

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @gabor-boros! Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @gabor-boros! Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @gabor-boros! Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.

File: tutor-contrib-harmony-plugin/tutor_k8s_harmony_plugin/harmony_search/base.py
Comment (mtyaka) [inline tutor-contrib-harmony-plugin/tutor_k8s_harmony_plugin/harmony_search/base.py:70]: Remove before merging

File: tutor-contrib-harmony-plugin/tutor_k8s_harmony_plugin/plugin.py
Comment (mtyaka) [inline tutor-contrib-harmony-plugin/tutor_k8s_harmony_plugin/plugin.py:112]:
```suggestion
        The default Elasticsearch configuration does not allow HTTP auth or CA
```

**Comment by cmltaWt0** [comment]
@gabor-boros Just want to mention that I'm going to test OpenSearch cluster. Sorry for such delay in doing the review.

File: README.md
Comment (cmltaWt0) [inline README.md:305]: We do not have a created namespace at this point yet. 
Should we mention this in the documentation?

**Comment by cmltaWt0** [comment]
Testing notes
---

I'm not able to get the `reindex_content_library` command work yet.
```
./manage.py lms reindex_content_library
...
File "/openedx/venv/lib/python3.8/site-packages/elasticsearch/connection/http_urllib3.py", line 263, in perform_request
    raise SSLError("N/A", str(e), e)
elasticsearch.exceptions.SSLError: ConnectionError([SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1131)) caused by: SSLError([SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1131))
```

However I can see the cert in django settings:
```
>>> settings.ELASTICSEARCH_CA_CERT
'-----BEGIN CERTIFICATE-----\XXXXXXX
XXXXXXXXXX
XXXXXXXXXX
XXXXXXXXXX\n-----END CERTIFICATE-----\n'
>>> settings.ELASTIC_SEARCH_INDEX_PREFIX
'openedx-01-einv-'
>>> settings.ELASTIC_SEARCH_CONFIG
[{'use_ssl': True, 'host': 'harmony-search-cluster.harmony.svc.cluster.local', 'port': 9200, 'http_auth': 'openedx-01:njN2pew2NfHALTQl3zii2cbK', 'ssl_context': <ssl.SSLContext object at 0xffffaf22b940>}]
```

Am I right that the current PR do not implement the mounting of the cert into forum? Maybe we have to update to testing instruction because it doesn't work for me following the instructions directly?
Forum obviously failing to verify the cert.
```
2024/06/12 12:24:13 Waiting for https://openedx-01:*************@harmony-search-cluster.harmony.svc.cluster.local:9200: Get "https://openedx-01:***@harmony-search-cluster.harmony.svc.cluster.local:9200": tls: failed to verify certificate: x509: certificate signed by unknown authority.
```


In general everything looks great, I can create a cluster, create an opensearch user, copy cert to INSTANCE_NAMESPACE and run everything. Just need to understand how to test it properly.

**Comment by gabor-boros** [comment]
@cmltaWt0 Are you using [this version](https://github.com/overhangio/tutor-forum/pull/35) of the forum (which is listed in the dependencies)? If yes, the forum pod should get the `ELASTICSEARCH_CA_PATH` as a docker build parameter from the `config.yaml` of Tutor and connect to the ES server accordingly. If that's not the case, please let me know as I may forgot something from the test instructions.

**Comment by cmltaWt0** [comment]
> @cmltaWt0 Are you using [this version](https://github.com/overhangio/tutor-forum/pull/35) of the forum (which is listed in the dependencies)? If yes, the forum pod should get the `ELASTICSEARCH_CA_PATH` as a docker build parameter from the `config.yaml` of Tutor and connect to the ES server accordingly. If that's not the case, please let me know as I may forgot something from the test instructions.

Yes. Using your branch:
```
❯ git -C tutor-forum branch
* gabor/wait-for-search-server-with-ca
```

And I have the volume in deployment spec:
```
spec:
      volumes:
        - name: search-cluster-certs
          secret:
            secretName: search-cluster-certificates
            defaultMode: 420
      containers:
        - name: forum
          image: docker.io/overhangio/openedx-forum:17.0.0
          ports:
            - containerPort: 4567
              protocol: TCP
          env:
            - name: SEARCH_SERVER
              value: >-
                https://openedx-01:njN2pew2NfHALTQl3zii2cbK@harmony-search-cluster.harmony.svc.cluster.local:9200
            - name: MONGODB_AUTH
            - name: MONGODB_HOST
              value: mongodb
            - name: MONGODB_PORT
              value: '27017'
            - name: MONGODB_DATABASE
              value: cs_comments_service
            - name: MONGOID_AUTH_SOURCE
              value: admin
            - name: MONGOID_AUTH_MECH
            - name: MONGOID_USE_SSL
              value: 'false'
            - name: ELASTICSEARCH_INDEX_PREFIX
              value: openedx-01-einv-
            - name: ELASTICSEARCH_CA_PATH
              value: /etc/ssl/certs/search-cluster.pem
          resources: {}
          volumeMounts:
            - name: search-cluster-certs
              mountPath: /etc/ssl/certs/search-cluster.pem
              subPath: ca.crt
          terminationMessagePath: /dev/termination-log
          terminationMessagePolicy: File
          imagePullPolicy: IfNotPresent
```
So we should be fine - `ca.crt` is mounted to `/etc/ssl/certs/search-cluster.pem`.
Need to investigate the cause of the error msg.


I will test it from scratch with Elasticsearch, not with OpenSearch.
So, am I correct in understanding that no additional actions are required beyond those described in the instructions?

**Comment by cmltaWt0** [comment]
@gabor-boros I've tested the flow from a scratch for ElasticSearch and it works perfectly!
It's good for me to merge this but add a note to README that work with OpenSearch is not confirmed yet. I'll re-test the OS separately and try to understand why it doesn't work for me.

**Comment by mphilbrick211** [comment]
> @gabor-boros I've tested the flow from a scratch for ElasticSearch and it works perfectly! It's good for me to merge this but add a note to README that work with OpenSearch is not confirmed yet. I'll re-test the OS separately and try to understand why it doesn't work for me.
> 
> P.S. Also we need to rebase the source branch on top of `main`.

@gabor-boros - friendly follow-up on this!

**Comment by gabor-boros** [comment]
@mphilbrick211 We are still waiting on a dependent tutor-forum PR. This PR is not forgotten

**Comment by gabor-boros** [comment]
@felipemontoya / @cmltaWt0 I would like to have another review/approve from you as I pushed some changes:

* https://github.com/openedx/openedx-k8s-harmony/pull/73/commits/601575d41227d546a751f6c062fddb31b5aeee8f
* https://github.com/openedx/openedx-k8s-harmony/pull/73/commits/9c49ab4b17233c6aa085bb0ec798c847845c15e2
* https://github.com/openedx/openedx-k8s-harmony/pull/73/commits/1a1d3a9b215161a91ed6c8705e00ca8b8c345962

File: tutor-contrib-harmony-plugin/tutor_k8s_harmony_plugin/plugin.py
Comment (cmltaWt0) [inline tutor-contrib-harmony-plugin/tutor_k8s_harmony_plugin/plugin.py:111]:
```suggestion
        Override forum env vars to configure the search cluster.
```

**Review by cmltaWt0 — APPROVED** [review]
Looks great. Just a one typo.

**Comment by openedx-webhooks** [comment]
@gabor-boros 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/openedx-k8s-harmony&entry.867055334=https://github.com/openedx/openedx-k8s-harmony/pull/73&entry.1484655318=https://github.com/gabor-boros&entry.752974735=2024-05-03+14:07&entry.1917517419=2024-07-31+09:21&entry.2133058324=Yes) so we can improve your experience in the future.

**Pull request merged**
